### PR TITLE
refactor(dev): disable resending pending messages [AR-2693]

### PR DESCRIPTION
**Pull request by saleniuk**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

Currently, if the message send fails, we schedule a worker to retry sending after some time or when the connection is back. At the current stage, we don't want to deal with such complex logic and just mark every message as failed to send when there's an error.

### Solutions

Refactor two use cases responsible for sending text and knock messages, which were using `sendPendingMessage` and scheduling the worker in case of failure. Now they try to send immediately and if it fails, the message is marked as failed. In the near future, we will implement the manual retry mechanism for the user to send such message again.

#### Test Coverage (Optional)

- [x] I have added automated test to this contribution

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/SendKnockUseCase.kt [hunk 1]
```diff
@@ -68,16 +71,29 @@ class SendKnockUseCase internal constructor(
...
             persistMessage(message)
-        }.flatMap {
-            messageSender.sendPendingMessage(conversationId, generatedMessageUuid)
+                .flatMap {
+                    messageSender.sendMessage(message)
+                }
+                .onSuccess {
+                    messageRepository.updateMessageStatus(MessageEntity.Status.SENT, conversationId, generatedMessageUuid)
+                }
         }.onFailure {
```
Comment (Garzas) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/SendKnockUseCase.kt:87]: I see that this handler is similar to SendEditTextMessageUseCase, can we somehow extract it to separate function?

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/SendTextMessageUseCase.kt [hunk 2]
```diff
@@ -92,9 +97,22 @@ class SendTextMessageUseCase internal constructor(
...
-        }.flatMap {
-            messageSender.sendPendingMessage(conversationId, generatedMessageUuid)
+                .flatMap {
+                    messageSender.sendMessage(message)
+                }
+                .onSuccess {
+                    messageRepository.updateMessageStatus(MessageEntity.Status.SENT, conversationId, generatedMessageUuid)
+                }
         }.onFailure {
+            when (it) {
```
Comment (Garzas) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/SendTextMessageUseCase.kt:107]: same here

**Review by Garzas — APPROVED** [review]
Nicely done ✅  Just suggestion to extract similar functionalities

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/SendKnockUseCase.kt (hunk 1 above)
Comment (saleniuk) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/SendKnockUseCase.kt]: Good idea, I moved the common code to the `MessageSendFailureHandler`.

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/MessageSendFailureHandler.kt [hunk 3]
```diff
@@ -34,11 +39,27 @@ interface MessageSendFailureHandler {
...
+
+    /**
+     * Handle a failure when attempting to send a message
+     * update the message status to FAILED or FAILED_REMOTELY depending on the resulted failure.
+     * @param failure the failure that occured
+     * @param conversationId id of the conversation of the message that failed
+     * @param messageId id of the message that failed
+     * @param messageType type of the message that failed (for logging purposes)
+     */
+    suspend fun handleFailureUpdateMessageStatus(
```
Comment (gongracr) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/MessageSendFailureHandler.kt:51]: at first look, I thought this method was handling an update event that had failed 🙈  
What about `handleFailedMessageDelivery` or `handleMessageFailureAndUpdateStatus` if we want to specify more concretely what action we are taking?

Naming is hard 😅

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/MessageSendFailureHandler.kt [hunk 4]
```diff
@@ -54,4 +75,30 @@ class MessageSendFailureHandlerImpl internal constructor(
...
+
+    override suspend fun handleFailureUpdateMessageStatus(
+        failure: CoreFailure,
+        conversationId: ConversationId,
+        messageId: String,
+        messageType: String
+    ) {
+        when (failure) {
+            is NetworkFailure.FederatedBackendFailure -> {
+                kaliumLogger.i("Sending message of type $messageType failed due to federation context availability.")
```
Comment (gongracr) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/MessageSendFailureHandler.kt:87]: is `i` verbose enough? or should we upgrade it to `w` or even `e`rror level?

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/MessageSendFailureHandler.kt [hunk 5]
```diff
@@ -54,4 +75,30 @@ class MessageSendFailureHandlerImpl internal constructor(
...
+            else -> {
+                messageRepository.updateMessageStatus(MessageEntity.Status.FAILED, conversationId, messageId)
+            }
+        }
+        if (failure is CoreFailure.Unknown) {
+            kaliumLogger.e(
+                "There was an unknown error trying to send the message of type: $messageType, cause: $failure",
+                failure.rootCause
+            )
+            failure.rootCause?.printStackTrace()
```
Comment (gongracr) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/MessageSendFailureHandler.kt:99]: why are we printing the error stack trace twice?

**Review by gongracr — APPROVED** [review]
Just tiny comments. Clean and spotless job once again ✨

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/MessageSendFailureHandler.kt (hunk 3 above)
Comment (saleniuk) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/MessageSendFailureHandler.kt]: You're right, I didn't know which name to choose, I changed it to `handleFailureAndUpdateMessageStatus`, thanks!

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/MessageSendFailureHandler.kt (hunk 4 above)
Comment (saleniuk) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/MessageSendFailureHandler.kt]: There's another log right after that so it was just to inform about that particular type, but I changed it to `e`

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/MessageSendFailureHandler.kt (hunk 5 above)
Comment (saleniuk) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/MessageSendFailureHandler.kt]: Right, I just moved the code without thinking much about it, but I think we don't want that, so I removed it 👍

**Pull request merged**